_posts/2019-02-11-configurable-builds-part-1.md

```diff
@@ -0,0 +1,284 @@
+---
+layout: posts
+title: Configurable Builds - Part 1
+authors:
+ - gregestren
+---
+
+One of Bazel's long-term goals is to make `$ bazel build //:all` "*just work*"
+so that every target builds "the right way", for whatever platform(s) you care
+about, using only flags that are genuinely interesting to you. This two-part
+series discusses the challenges involved and steps we're taking to get there.
+
+This is a deeper dive into themes covered in Bazel's [configurability roadmap]
+(https://bazel.build/roadmaps/configuration.html).
+
+## Motivation
+
+The larger your project gets, the more likely you are to have to build it
+different ways.
+
+Maybe you need Android, iOS, and desktop versions of your app. Maybe you build
+C++ for different platforms. Or maybe you maintain a popular JavaScript library
+that users only want parts of to keep their .js files small.
+
+These are examples of *configuration*: the process of building the same *code*
+with different *settings* to customize it for specific needs. In Bazel-speak,
+this means building a
+[target](https://docs.bazel.build/versions/master/build-ref.html#targets) with
+a set of
+[flags](https://docs.bazel.build/versions/master/user-manual.html#flags-options):
+
+```sh
+$ bazel build //my:cc_binary --cpu=x86
+$ bazel build //my:cc_binary --cpu=arm
+$ bazel build //my:android_binary --android_sdk=@androidsdk//:sdk-28
+```
+
+Bazel's *configurability work* is an ongoing effort to make these tasks simple,
+flexible, and powerful. 
+
+## The Problem
+
+Bazel is a
+[powerful](https://docs.bazel.build/versions/master/bazel-vision.html) build
+tool that's especially suited for large codebases with
+[multiple languages](https://blog.bazel.build/2018/12/05/multilanguage-build-system.html).
+But it grew out of Google, which historically wrote code for fleets of identical
+Linux servers with little need for customization. What customization was needed
+was achieved with
+[ad hoc flags and logic](https://source.bazel.build/bazel/+/144912e7b7a86b45e07f79e76f6fed20890acb36:src/main/java/com/google/devtools/build/lib/rules/cpp/CppOptions.java;l=258)
+built straight into the tool.
+
+
+This is no longer true inside or outside of Google. Modern software targets
+phones, cloud, servers, desktops, [smart
+devices](https://en.wikipedia.org/wiki/Smart_device) and more, and must offer
+increasingly flexible support for any combination of these platforms.
+
+This means
+
+    $ bazel build //my:binary
+
+isn't enough to describe what you want to do. What platform(s) are you
+targeting? What features do you want to include? What if you want to build a
+complex app with generated sources, client and server modules, native
+extensions, and test data?
+
+Bazel's historical approach of ad hoc flags:
```

Comment:
Rather than splitting the sentence with code, which is hard to read, what about completing both sentences? E.g., "Bazel's historical approach is to use ad-hoc flags:" and "This approach has many limitations:" (or "This approach is lacking in many ways.").

Reply:
Done. This definitely works.